python/ray/train/v2/_internal/execution/controller/state.py

```diff
@@ -0,0 +1,142 @@
+        is_active() -> bool: Returns whether the state is active.
+        needs_new_run_attempt() -> bool: Returns whether a new run attempt is needed.
+    """
+
+    def __init__(self, state_type: TrainControllerStateType):
+        self._state_type = state_type
+
+    def __repr__(self) -> str:
+        attrs = {
+            "type": self._state_type.name,
+            "is_active": self._state_type.is_active,
+            "needs_new_run_attempt": self._state_type.needs_new_run_attempt,
+            **{k: v for k, v in vars(self).items() if not k.startswith("_")},
+        }
+        attrs_str = "\n    ".join(f"{k}={v}" for k, v in attrs.items())
+        return f"{self.__class__.__name__}(\n    {attrs_str}\n)"
+
+    def is_active(self) -> bool:
+        return self._state_type.is_active
+
+    def needs_new_run_attempt(self) -> bool:
+        return self._state_type.needs_new_run_attempt
+
+
+class InitializingState(TrainControllerState):
+    def __init__(self):
+        super().__init__(state_type=TrainControllerStateType.INITIALIZING)
+
+
+class SchedulingState(TrainControllerState):
+    def __init__(self, scaling_decision: ScalingDecision):
+        super().__init__(state_type=TrainControllerStateType.SCHEDULING)
+        self.scaling_decision = scaling_decision
+
+
+class ReschedulingState(TrainControllerState):
+    def __init__(self):
+        super().__init__(state_type=TrainControllerStateType.RESCHEDULING)
+
+
+class RunningState(TrainControllerState):
+    # TODO: Split into multiple states where each state has exact non-optional fields.
+    def __init__(
+        self,
+        worker_group_status: Optional[WorkerGroupStatus] = None,
+        failure_decision: Optional[FailureDecision] = None,
+    ):
+        super().__init__(state_type=TrainControllerStateType.RUNNING)
+        self.worker_group_status = worker_group_status
+        self.failure_decision = failure_decision
+
+
+class RestartingState(TrainControllerState):
+    def __init__(
+        self,
+        worker_group_status: WorkerGroupStatus,
+        failure_decision: FailureDecision,
+        training_failed_error: TrainingFailedError,
+    ):
+        super().__init__(state_type=TrainControllerStateType.RESTARTING)
+        self.worker_group_status = worker_group_status
+        self.failure_decision = failure_decision
+        self.training_failed_error = training_failed_error
+
+
+class ResizingState(TrainControllerState):
+    def __init__(
+        self, worker_group_status: WorkerGroupStatus, scaling_decision: ScalingDecision
+    ):
+        super().__init__(state_type=TrainControllerStateType.RESIZING)
+        self.worker_group_status = worker_group_status
+        self.scaling_decision = scaling_decision
+
+
+class ErroredState(TrainControllerState):
+    def __init__(
+        self,
+        worker_group_status: WorkerGroupStatus,
+        failure_decision: FailureDecision,
```

Comment:
Doesn't need these two

Reply:
They're not used right now but I was thinking it might be helpful for logging/debuggability, in particular propagating the `FailureDecision` to anything that might need to process it. 

I'm also okay to remove for now and add it back in later when there's a need for it.